src/Raven.Server/Documents/Handlers/BatchRequestParser.cs

```diff
@@ -124,12 +128,25 @@ private static void ReturnBuffer(CommandData[] cmds)
 
                     if (commandData.Type == CommandType.PUT && string.IsNullOrEmpty(commandData.Id) == false && commandData.Id[commandData.Id.Length - 1] == '|')
                     {
-                        var (_, id, _) = await serverStore.GenerateClusterIdentityAsync(commandData.Id, database.Name);
-                        commandData.Id = id;
+                        // queue identities requests in order to send them at once to the leader (using List for simplicity)
+                        identities.Add(commandData.Id);
                     }
 
                     cmds[index] = commandData;
                 }
+
+                var newIds = await serverStore.GenerateClusterIdentitiesBatchAsync(database.Name, identities);
+                Debug.Assert(newIds.Count == identities.Count);
+
+                for (var i = 0; i < cmds.Length; i++)
+                {
+                    if (cmds[i].Type == CommandType.PUT && string.IsNullOrEmpty(cmds[i].Id) == false && cmds[i].Id[cmds[i].Id.Length - 1] == '|')
```

Comment:
Here you can create a dictionary of the position in the list to the command in question